service/src/main.rs

```diff
@@ -529,51 +529,59 @@ pub fn init_light_client<E: EnclaveBase + SideChain>(
 	produce_blocks(enclave_api, api, latest)
 }
 
-/// Starts block production
+/// Gets the amount of blocks to sync from the parentchain and feeds them to the enclave.
 ///
-/// Returns the last synced header of layer one
+/// It also calls into the enclave if there are no blocks to sync because the enclave does
+/// currently two things in `produce_blocks`:
+///
+/// * sync blocks
+/// * execute pending trusted operations
+///
+/// Todo: the two tasks above should be independent: #404
 pub fn produce_blocks<E: EnclaveBase + SideChain>(
 	enclave_api: &E,
 	api: &Api<sr25519::Pair, WsRpcClient>,
-	last_synced_head: Header,
+	mut last_synced_header: Header,
 ) -> Header {
-	// obtain latest finalized block from layer one
-	debug!("Getting current head");
+	let tee_accountid = enclave_account(enclave_api);
+
+	trace!("Getting current head");
 	let curr_head: SignedBlock = api.last_finalized_block().unwrap().unwrap();
+	let head_block_number = curr_head.block.header.number;
 
-	let blocks_to_sync = get_blocks_to_sync(api, &last_synced_head, &curr_head);
+	let blocks_to_sync = get_blocks_to_sync(api, &last_synced_header, &curr_head);
 
-	let tee_accountid = enclave_account(enclave_api);
+	if blocks_to_sync.is_empty() {
+		// we have nothing to sync, but we can still execute trusted operations
+		let tee_nonce = api.get_nonce_of(&tee_accountid).unwrap();
 
-	// only feed BLOCK_SYNC_BATCH_SIZE blocks at a time into the enclave to save enclave state regularly
-	let mut current_block_number = if curr_head.block.header.hash() == last_synced_head.hash() {
-		curr_head.block.header.number as usize
-	} else {
-		blocks_to_sync.first().map(|h| h.block.header.number as usize).unwrap_or(0)
-	};
+		if let Err(e) = enclave_api.produce_blocks::<Block>(&[], tee_nonce) {
+			error!("{:?}", e);
+		};
 
+		return last_synced_header
+	}
```

Comment:
So this is the key part? We now check if there are not blocks not be synced, and still call into the enclave  with `produce_blocks`? Because `produce_blocks` does more than just produce blocks, but also execute tops?

Reply:
Exactly. Unfortunately it is like that.

Reply:
Well, I did not introduce the naming, but I think `produce_blocks` got its name because it produces sidechain blocks, but it does, however, also sync parentchain blocks. But before changing the code here too heavily, I wanted to think about designs for #400, and #247.

Reply:
Thanks for confirming 👍 I mainly asked to make sure I understand what is going on in this PR 😄 - as for improving the situation, you've already opened new issues, great!

Reply:
Regardless, you are right. I can still do a proper renaming that explains a bit better what is happening inside.